[WIP] Cleanup API design and implementation

rpjohnst opened the pull request:
Just rearranges the API types, just not verify them yet. Wanted to let someone take a look at it first.

Closes #54

rpjohnst committed:
Cleanup API design and implementation

- move discriminants inside `data` fields
- use Rust enums instead of `serde_json::Value`
- simplify some API types
- does not yet verify incoming types

elliothatch commented on client/app/models/Errors.ts:
```diff
@@ -0,0 +1,23 @@
+export class InvalidTypeError extends Error {
+	constructor(typeName: string) {
+		super(`Expected to get type ${typeName} in this context`);
```
Why did you remove the data field? It was really useful for debugging.

elliothatch reviewed (approved):
Looks good overall, but I'd really like to have data validation before we merge. It saves a lot of time debugging our own code when there are protocol contract violations.

rpjohnst committed:
Verify JSON types from the server

- a @Deserialize decorator marks required fields; fromJSON does the checks
  (variant data is not checked as TypeScript doesn't emit the necessary
  metadata to check interfaces)
- output es6, since this is really an app there should be no problems
  (we can use Babel if there are; TypeScript is insufficient)
- bring back some Error type data fields
- since they're classes again, may as well make getParametersAsString a
  method again

rpjohnst committed:
Merge branch 'master' into api-cleanup

# Conflicts:
#	client/app/components/debugger/debugger.component.ts
#	client/app/components/functions/functions.component.ts
